### PR TITLE
feat(sync): insert history-lost systemMessage after 404 sync recovery (AR-2897)

### PR DESCRIPTION
**Pull request by mchenani**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues
After the 404 error in sync, since the would happen if the user stays offline for long time, there is a chance of missing messages and losing history!
We need to inform the user about that by inserting a local system message for all conversations that indicates the history lost probability. 

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution
----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/MessageRepository.kt [hunk 1]
```diff
@@ -211,6 +215,14 @@ class MessageDataSource(
         )
     }
 
+    override suspend fun persistSystemMessageToAllConversations(
+        message: Message.System
+    ): Either<CoreFailure, Unit> = wrapStorageRequest {
+        messageMapper.fromMessageToEntity(message).let {
+            if (it is MessageEntity.System) messageDAO.persistSystemMessageToAllConversations(it)
```
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/MessageRepository.kt:222]: Should this return an `OnlySystemMessageAllowed` in the else case?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/incremental/IncrementalSyncRecoveryHandler.kt [hunk 2]
```diff
@@ -16,14 +17,16 @@ internal fun interface OnIncrementalSyncRetryCallback {
...
+    private val slowSyncRepository: SlowSyncRepository,
+    private val addSystemMessageToAllConversationsUseCase: AddSystemMessageToAllConversationsUseCase
 ) : IncrementalSyncRecoveryHandler {
 
     override suspend fun recover(failure: CoreFailure, onIncrementalSyncRetryCallback: OnIncrementalSyncRetryCallback) {
         kaliumLogger.i("$TAG Checking if we can recover from the failure: $failure")
         if (shouldRestartSlowSyncProcess(failure)) {
             slowSyncRepository.clearLastSlowSyncCompletionInstant()
             slowSyncRepository.setNeedsToRecoverMLSGroups(true)
+            addSystemMessageToAllConversationsUseCase()
```
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/incremental/IncrementalSyncRecoveryHandler.kt:29]: Let's add unit test checking that this use case is invoked?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/MessageRepository.kt [hunk 3]
```diff
@@ -211,6 +215,14 @@ class MessageDataSource(
         )
     }
 
+    override suspend fun persistSystemMessageToAllConversations(
```
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/MessageRepository.kt:218]: Let's add unit test checking the message really inserted into all conversations?

File: persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq [hunk 4]
```diff
@@ -418,6 +418,10 @@ insertOrIgnoreMessage:
 INSERT OR IGNORE INTO Message(id, content_type, conversation_id, date, sender_user_id, sender_client_id, status, visibility, expects_read_confirmation)
 VALUES(?, ?, ?, ?, ?, ?, ?, ?, ?);
 
+insertOrIgnoreBullkSystemMessage:
```
Comment (typfel) [inline persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq:421]:
```suggestion
insertOrIgnoreBulkSystemMessage:
```

File: persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq [hunk 5]
```diff
@@ -418,6 +418,10 @@ insertOrIgnoreMessage:
...
+INSERT OR IGNORE INTO Message(id, content_type, conversation_id, date, sender_user_id, sender_client_id, status, visibility, expects_read_confirmation)
```
Comment (typfel) [inline persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq:422]: Should maybe limit the insertion to only 1:1 conversations and groups?

**Review by typfel — CHANGES_REQUESTED** [review]
Let's add some tests and I added some questions/suggestions.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/MessageRepository.kt (hunk 1 above)
Comment (mchenani) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/MessageRepository.kt:222]: I added the error for it, but missed it in the final step here :D thanks for mentioning it.

File: persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq (hunk 5 above)
Comment (mchenani) [reply persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq:422]: Hmm - ok! I'll do it but still no harm of adding the message for all.

**Review by typfel — APPROVED** [review]

**Pull request merged**